firedrake/slate/preconditioners.py

```diff
@@ -251,22 +276,35 @@ def apply(self, pc, x, y):
         Lastly, we project the broken solutions into the mimetic
         non-broken finite element space.
         """
-        with self.unbroken_rhs.dat.vec as v:
+        with self.unbroken_residual.dat.vec as v:
             x.copy(v)
 
         # Transfer unbroken_rhs into broken_rhs
-        unbroken_scalar_data = self.unbroken_rhs.split()[self.pidx]
-        broken_scalar_data = self.broken_rhs.split()[self.pidx]
-        self.data_projector.project()
+        # NOTE: Scalar space is already "broken" so no need for
+        # any projections
+        unbroken_scalar_data = self.unbroken_residual.split()[self.pidx]
+        broken_scalar_data = self.broken_residual.split()[self.pidx]
         unbroken_scalar_data.dat.copy(broken_scalar_data.dat)
 
+        # Handle the incoming HDiv residual:
+        # Solve (approximately) for `g = A.inv * r`, where `A` is the HDiv
+        # mass matrix and `r` is the HDiv residual.
+        # Once `g` is obtained, we take the inner product against broken
+        # HDiv test functions to obtain a broken residual.
+        with self.unbroken_residual.split()[self.vidx].dat.vec_ro as hdr:
+            with self._computed_r_map.dat.vec as g:
+                self.hd_ksp.solve(hdr, g)
+
+        # Now assemble the new "broken" hdiv residual
+        self._assemble_broken_hdr()
+
         # Compute the rhs for the multiplier system
         self._assemble_Srhs()
 
         # Solve the system for the Lagrange multipliers
         with self.schur_rhs.dat.vec_ro as b:
-            with self.trace_solution.dat.vec as x:
-                self.ksp.solve(b, x)
+            with self.trace_solution.dat.vec as x_trace:
+                self.trace_ksp.solve(b, x_trace)
```

Comment:
You should add viewer calls to the other configurable objects: to whit the forward dual->primal KSP and the reconstruction project call.